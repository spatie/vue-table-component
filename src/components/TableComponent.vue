<template>
    <div class="table-component">
        <div v-if="showFilter && filterableColumnExists" class="table-component__filter">
            <input
                    :class="fullFilterInputClass"
                    type="text"
                    v-model="filter"
                    :placeholder="filterPlaceholder"
            >
            <a
                    v-if="filter"
                    @click="filter = ''"
                    class="table-component__filter__clear"
            >Ã—</a>
        </div>

        <div class="table-component__table-wrapper">
            <table :class="fullTableClass">
                <caption v-if="showCaption" class="table-component__table__caption" role="alert" aria-live="polite">
                    {{ ariaCaption }}
                </caption>
                <thead :class="fullTableHeadClass">
                <tr>
                    <table-column-header
                            @click="changeSorting"
                            v-for="column in columns"
                            :key="column.show"
                            :sort="sort"
                            :column="column"
                    ></table-column-header>
                </tr>
                </thead>
                <tbody :class="fullTableBodyClass">
                <table-row
                        v-for="row in displayedRows"
                        :key="row.vueTableComponentInternalRowId"
                        :row="row"
                        :columns="columns"
						:class="fullTableRowClass(row)"
						@rowClick="emitRowClick"
                ></table-row>
                </tbody>
                <tfoot>
                    <slot name="tfoot" :rows="rows"></slot>
                </tfoot>
            </table>
        </div>

        <div v-if="displayedRows.length === 0" class="table-component__message">
            {{ filterNoResults }}
        </div>


        <div style="display:none;">
            <slot></slot>
        </div>

        <pagination v-if="pagination" :pagination="pagination" @pageChange="pageChange"></pagination>
    </div>
</template>

<script>
    import Column from '../classes/Column';
    import expiringStorage from '../expiring-storage';
    import Row from '../classes/Row';
    import TableColumnHeader from './TableColumnHeader';
    import TableRow from './TableRow';
    import settings from '../settings';
    import Pagination from './Pagination';
    import { classList, pick } from '../helpers';

    export default {
        components: {
            TableColumnHeader,
            TableRow,
            Pagination,
        },

        props: {
            data: { default: () => [], type: [Array, Function] },

            showFilter: { default: true },
            showCaption: { default: true },

            sortBy: { default: '', type: String },
            sortOrder: { default: '', type: String },

            cacheKey: { default: null },
            cacheLifetime: { default: 5 },

            tableClass: { default: settings.tableClass },
            theadClass: { default: settings.theadClass },
			tbodyClass: { default: settings.tbodyClass },
			rowClass:  { default: settings.rowClass },
            filterInputClass: { default: settings.filterInputClass },
            filterPlaceholder: { default: settings.filterPlaceholder },
            filterNoResults: { default: settings.filterNoResults },
        },

        data: () => ({
            columns: [],
            rows: [],
            filter: '',
            sort: {
                fieldName: '',
                order: '',
            },
            pagination: null,

            localSettings: {},
        }),

        created() {
            this.sort.fieldName = this.sortBy;
            this.sort.order = this.sortOrder;

            this.restoreState();
        },

        async mounted() {
            const columnComponents = this.$slots.default
                .filter(column => column.componentInstance)
                .map(column => column.componentInstance);

            this.columns = columnComponents.map(
                column => new Column(column)
            );

            columnComponents.forEach(columnCom => {
                Object.keys(columnCom.$options.props).forEach(
                    prop => columnCom.$watch(prop, () => {
                        this.columns = columnComponents.map(
                            column => new Column(column)
                        );
                    })
                );
            });

            await this.mapDataToRows();
        },

        watch: {
            filter() {
                if (!this.usesLocalData) {
                    this.mapDataToRows();
                }

                this.saveState();
            },

            data() {
                if (this.usesLocalData) {
                    this.mapDataToRows();
                }
            },
        },

        computed: {
            fullTableClass() {
                return classList('table-component__table', this.tableClass);
            },

            fullTableHeadClass() {
                return classList('table-component__table__head', this.theadClass);
            },

            fullTableBodyClass() {
                return classList('table-component__table__body', this.tbodyClass);
            },

            fullFilterInputClass() {
                return classList('table-component__filter__field', this.filterInputClass);
            },

            ariaCaption() {
                if (this.sort.fieldName === '') {
                    return 'Table not sorted';
                }

                return `Table sorted by ${this.sort.fieldName} ` +
                    (this.sort.order === 'asc' ? '(ascending)' : '(descending)');
            },

            usesLocalData() {
                return Array.isArray(this.data);
            },

            displayedRows() {
                if (!this.usesLocalData) {
                    return this.sortedRows;
                }

                if (!this.showFilter) {
                    return this.sortedRows;
                }

                if (!this.columns.filter(column => column.isFilterable()).length) {
                    return this.sortedRows;
                }

                return this.sortedRows.filter(row => row.passesFilter(this.filter));
            },

            sortedRows() {
                if (!this.usesLocalData) {
                    return this.rows;
                }

                if (this.sort.fieldName === '') {
                    return this.rows;
                }

                if (this.columns.length === 0) {
                    return this.rows;
                }

                const sortColumn = this.getColumn(this.sort.fieldName);

                if (!sortColumn) {
                    return this.rows;
                }

                return this.rows.sort(sortColumn.getSortPredicate(this.sort.order, this.columns));
            },

            filterableColumnExists() {
                return this.columns.filter(c => c.isFilterable()).length > 0;
            },

            storageKey() {
                return this.cacheKey
                    ? `vue-table-component.${this.cacheKey}`
                    : `vue-table-component.${window.location.host}${window.location.pathname}${this.cacheKey}`;
            },
        },

        methods: {
            async pageChange(page) {
                this.pagination.currentPage = page;

                await this.mapDataToRows();
            },

            async mapDataToRows() {
                const data = this.usesLocalData
                    ? this.prepareLocalData()
                    : await this.fetchServerData();

                let rowId = 0;

                this.rows = data
                    .map(rowData => {
                        rowData.vueTableComponentInternalRowId = rowId++;
                        return rowData;
                    })
                    .map(rowData => new Row(rowData, this.columns));
            },

            prepareLocalData() {
                this.pagination = null;

                return this.data;
            },

            async fetchServerData() {
                const page = this.pagination && this.pagination.currentPage || 1;

                const response = await this.data({
                    filter: this.filter,
                    sort: this.sort,
                    page: page,
                });

                this.pagination = response.pagination;

                return response.data;
            },

            async refresh() {
                await this.mapDataToRows();
            },

            changeSorting(column) {
                if (this.sort.fieldName !== column.show) {
                    this.sort.fieldName = column.show;
                    this.sort.order = 'asc';
                } else {
                    this.sort.order = (this.sort.order === 'asc' ? 'desc' : 'asc');
                }

                if (!this.usesLocalData) {
                    this.mapDataToRows();
                }

                this.saveState();
            },

            getColumn(columnName) {
                return this.columns.find(column => column.show === columnName);
            },

            saveState() {
                expiringStorage.set(this.storageKey, pick(this.$data, ['filter', 'sort']), this.cacheLifetime);
            },

            restoreState() {
                const previousState = expiringStorage.get(this.storageKey);

                if (previousState === null) {
                    return;
                }

                this.sort = previousState.sort;
                this.filter = previousState.filter;

                this.saveState();
            },

			emitRowClick(row) {
				this.$emit('rowClick', row);
			},

            fullTableRowClass(row) {
				const classResult = typeof this.rowClass === 'function' ? `${this.rowClass(row)}` : this.rowClass
                return classList('table-component__table__body__row', classResult);
            },
		},
    };
</script>
