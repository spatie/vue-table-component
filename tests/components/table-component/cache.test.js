import TableComponent from '../../../src';
import Vue from 'vue/dist/vue.js';
import expiringStorage from '../../../src/expiring-storage';
import simulant from 'simulant';

import LocalStorageMock from '../../helpers/LocalStorageMock';

const localStorage = new LocalStorageMock();

window.localStorage = localStorage;

describe('Caching tableComponent', () => {
    Vue.use(TableComponent);

    beforeEach(() => {
        localStorage.clear();

        document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        cache-key="test"
                        :data="[{ id: 1, firstName: 'John', lastName: 'Lennon' },
                                { id: 2, firstName: 'Paul', lastName: 'McCartney' }]"
                    >
                        <table-column show="firstName" label="First name"></table-column>
                        <table-column show="lastName" label="Last name"></table-column>
                    </table-component>
                </div>
            </div>
        `;

        const dateClass = Date;

        // eslint-disable-next-line no-global-assign
        Date = function (dateString) {
            return new dateClass(dateString || '2017-01-01T00:00:00.000Z');
        };
    });

    it('will cache the used filter and sorting', async () => {
        const cacheContent = {
            filter: 'Paul',
            sort:{
                fieldName: 'firstName',
                order:'asc',
            },
        };

        expiringStorage.set('vue-table-component.test', cacheContent, 5);

        progressTime(4);

        const table = await createVm();

        expect(table.filter).toEqual('Paul');
        expect(table.sort.fieldName).toEqual('firstName');
        expect(table.sort.order).toEqual('asc');
    });

    it('will not use the cache when it has expired', async () => {
        const cacheContent = {
            filter: 'Paul',
            sort:{
                fieldName: 'firstName',
                order:'asc',
            },
        };

        expiringStorage.set('vue-table-component.test', cacheContent, 5);

        progressTime(6);

        const table = await createVm();

        expect(table.filter).toEqual('');
        expect(table.sort.fieldName).toEqual('');
        expect(table.sort.order).toEqual('');
    });

    it('will cache the filter', async () => {
        const table = await createVm();

        table.filter = 'cache this';

        await Vue.nextTick(() => {});

        expect(localStorage.getAll()).toMatchSnapshot();
    });

    it('will cache a the sort column', async () => {
        document.body.innerHTML = `
            <div id="app">
                <div>
                    <table-component
                        cache-key="test"
                        :data="[{ firstName: 'John', songs: 30 },
                                { firstName: 'Paul', songs: 20 },
                                { firstName: 'George', songs: 420 },
                                { firstName: 'Ringo', songs: 210 }]"
                    >
                        <table-column show="firstName" label="First name" sort-by="songs"></table-column>
                        <table-column show="songs" data-type="numeric" label="Songs" sort-by="songs"></table-column>
                    </table-component>
                </div>
            </div>
        `;

        await createVm();

        const secondColumnHeader = document.getElementsByTagName('th')[1];

        await simulant.fire(secondColumnHeader, 'click');

        expect(localStorage.getAll()).toMatchSnapshot();
    });
});

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {});

    const table = vm.$children[0];

    return table;
}

function progressTime(minutes) {
    const currentTime = (new Date()).getTime();

    const newTime = new Date(currentTime + (minutes * 60000));

    const originalDateClass = Date;

    // eslint-disable-next-line no-global-assign
    Date = function (dateString) {
        return new originalDateClass(dateString || newTime.toISOString());
    };
}
