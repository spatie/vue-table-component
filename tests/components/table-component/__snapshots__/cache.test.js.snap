// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Caching tableComponent will cache a the sort column 1`] = `
Object {
  "vue-table-component.test": "{\\"value\\":{\\"filter\\":\\"\\",\\"sort\\":{\\"fieldName\\":\\"songs\\",\\"order\\":\\"asc\\"}},\\"expires\\":\\"2017-01-01T00:05:00.000Z\\"}",
}
`;

exports[`Caching tableComponent will cache the filter 1`] = `
Object {
  "vue-table-component.test": "{\\"value\\":{\\"filter\\":\\"cache this\\",\\"sort\\":{\\"fieldName\\":\\"\\",\\"order\\":\\"\\"}},\\"expires\\":\\"2017-01-01T00:05:00.000Z\\"}",
}
`;
